File: 402-remove-k-digits/remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = deque()
        for digit in num:
            while k > 0 and stack and digit < stack[-1]:
                stack.pop()
                k -= 1
            stack.append(digit)
        
        for _ in range(k):
            stack.pop()
        
        while stack and stack[0] == '0':
            stack.popleft()
        
        return ''.join(stack) if stack else '0'

File: 402-remove-k-digits/remove-k-digits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */

var removeKdigits = function (num, k) {
    const stack = [];
    let remainingToRemove = k;

    for (let digit of num) {
        while (remainingToRemove > 0 && stack.length > 0 && digit < stack[stack.length - 1]) {
            stack.pop();
            remainingToRemove--;
        }
        stack.push(digit);
    }

    while (remainingToRemove > 0) {
        stack.pop();
        remainingToRemove--;
    }

    let result = stack.join('').replace(/^0+/, '');

    if (result === '') return '0';
    return result;
};